rfcs/20200920-keras-nlp-bert.md

```diff
@@ -0,0 +1,370 @@
+# keras-nlp Transformer Encoder API
+
+| Status        | Proposed      |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | Zhenyu Tan (tanzheny@google.com), Francois Chollet (fchollet@google.com), Mark Omernick (momernick@google.com), Hongkun Yu (hongkuny@google.com)|
+| **Updated**   | 2020-09-21                                           |
+
+
+## Objective
+
+We aim at providing a set of primitive Keras layers to handle Transformer Encoder based, BERT-style models.
+
+Tokenization is not part of thi 
+
+## Key Benefits
+
+BERT-style transformer encoders has been state-of-art technique that enpowers many NLP tasks:
+- single sentence classification task, e.g., sentiment analysis
+- sentence pair classification task, e.g., next sentence prediction
+- question answering task, e.g., SQuAD
+- single sentence tagging task, e.g., named entity recognition
+
+With this proposal, users of will be able to model above language tasks with much simplified API. 
+
+## Design overview
+
+This proposal builds on the assumption that inputs are lookup indices, i.e., tf.int64. Tokenization is not part of this
+proposal but will be our immediate next step.
+
+### Classification task
+
+Case where a user want to use a pretrained BERT encoder for sentiment analysis:
+```python
+# Considering a imbd review dataset
+import tensorflow as tf
+import tensorflow_datasets as tfds
+import keras_nlp
+import tensorflow_text as tftext
+
+imdb_reviews = tfds.load('imdb_reviews')
+train_ds = imdb_reviews['train'].batch(32)
+test_ds = imdb_reviews['test'].batch(32)
+
+# Tokenization with BertTokenizer
+vocab_path = "gs://<bucket_name>/<file_path>/vocab.txt"
+tokenizer = tftext.BertTokenizer(vocab_path, token_out_type=tf.int64, lower_case=False)
+MAX_SEQ_LENGTH = 128
+def preprocess(input_text):
+  token_ids = tokenizer.tokenize(input_text)
+  cls = tf.expand_dims(tf.expand_dims(tf.cast(tf.tile([101], [token_ids.nrows()]), tf.int64), axis=1), axis=1)
+  sep = tf.expand_dims(tf.expand_dims(tf.cast(tf.tile([102], [token_ids.nrows()]), tf.int64), axis=1), axis=1)
+  token_ids = tf.concat([cls, token_ids, sep], axis=1)
+  segment_ids = tf.concat([tf.zeros_like(cls), tf.ones_like(token_ids), tf.ones_like(sep)], axis=1)
+  output_shape = [None, SEQUENCE_LENGTH]
```

Comment:
Where is SEQUENCE_LENGTH defined? It's MAX_SEQ_LENGTH?

Reply:
Fixed.